Interactive Problem from Julian & Finn (#20)
compprog-lecture-tools/fun-contest-2023: cp23/hpi-fun2/maze/problem.tex

\input{template.tex}

\begin{document}

\makeheader

One night Alice wakes up in a dark room. She has read about this place in one of her books. It is a large maze that is known to have exactly one exit in it. But Alice is running out of time. If she doesn't find her way to the exit soon, she will be trapped in it forever.

The only thing that Alice knows about this place is that it is build in a grid style. Each room in the grid is either a room or completely blocked. She can only travel to one of the four rooms directly adjacent to her current room.

Since it is very dark, she can only feel in the room she is currently in, if the adjacent rooms are open or blocked. Can you help Alice find a way out of the maze before she will get trapped there forever?

\paragraph*{Interaction}

This is an interactive problem. This means that your program will not first read all intput and then write some output, but instead communicates with the jury by writing queries to standard output and reading the answers from standard input.

> In the first line, you recieve two integer values $w$ and $h$ ($3\leq w, h \leq 10^9$), the width and height of the maze.

Reviewer: You should probably bound the product of w and h because as is the problem isn't always solvable.

Reviewer: Actually, it's possible to solve the problem without being given w and h, so you might consider removing this if you want to. It's also fine to leave as is though.

Author: We also thought about this (leaving out w and h) but then we couldn't give a max number of interactions and to solve the problem one would need to use a map instead of a vector.

To you think this problem would be to easy if we provide `w` and `h`?

Reviewer: Oh yeah, that's true. So I think giving them is fine. You should still bound the product though.

Author: I just changed to problem statement (and decided to bound it to `w * h <= 10^6`. Not sure if we should go even higher since runtime is `O(4* w* h)`.

Reviewer: I wouldn't go any higher, actually you probably need to lower it a bit further because interaction and flushing is pretty slow. That's something you'll figure out through testing.

Author: But how can we test this? It is mentioned in the Repo, that interactive problems aren't supported for automated testing :(

As this problem is interactive, you need to flush \texttt{stdout} after each operation, because otherwise Alice won't be able to receive your move request.
\begin{description}
	\itemsep-0.4em
	\item[Java:] \texttt{System.out.flush()}
	\item[C++:] \texttt{endl} or \texttt{cout.flush()}
  \item[Python] \texttt{print()} flushes per default
\end{description}

\paragraph*{Output}

Print the abreviation of one of the four directions \texttt{N}, \texttt{E}, \texttt{S}, \texttt{W} to move Alice in this direction. If you send her in a direction where the room is blocked, she will be mad at you and respond with \texttt{-1}.

If Alice reaches the exit of the maze, she will print \texttt{EXIT}. Otherwise, she will look around in the room telling you for each of the directions if there is a room availalbe or not.

Note: It is always guaranteed that a valid path to the exit exists and the maze will not change. However, there might be dead ends and loops within the maze.

\paragraph*{Sample interaction}

> The maze in this sample usecase looks like this (it is still surrounded by a wall that is not part of the maze) where $s$ represents Alice position at the beginning and $x$ marks the exit.

Reviewer: I'd write testcase rather than usecase

\input{template.tex}

\begin{document}

\makeheader

One night Alice wakes up in a dark room. She has read about this place in one of her books. It is a large maze that is known to have exactly one exit in it. But Alice is running out of time. If she doesn't find her way to the exit soon, she will be trapped in it forever.

The only thing that Alice knows about this place is that it is build in a grid style. Each room in the grid is either a room or completely blocked. She can only travel to one of the four rooms directly adjacent to her current room.

> Since it is very dark, she can only feel in the room she is currently in, if the adjacent rooms are open or blocked. Can you help Alice find a way out of the maze before she will get trapped there forever?

Reviewer: ```suggestion
Since it is very dark, she can only feel in the room she is currently in, if the adjacent rooms are open or blocked. Can you help Alice find a way out of the maze before she gets trapped forever?
```

\paragraph*{Interaction}

This is an interactive problem. This means that your program will not first read all input and then write some output, but instead communicates with the jury by writing queries to standard output and reading the answers from standard input.

> In the first line, you are given two integers $w$ and $h$ ($2\leq w \cdot h \leq 5 \cdot 10^5$), the width and height of the maze.

Reviewer: I think it's more clear (and prevents $w,h < 0$) if you individually bound them as well.

Reviewer: I would lower the bound to $2 \cdot 10^5$.
